File: problems/1254_number_of_closed_islands/Solution.java
class Solution {
    int[][] grid;
    int m, n;
    public int closedIsland(int[][] grid) {
        this.grid = grid; m = grid.length; n = grid[0].length;
        int count = 0;
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++)
                if (grid[i][j] == 0 && dfs(i, j)) count++;
        return count;
    }
    boolean dfs(int r, int c) {
        if (r < 0 || r >= m || c < 0 || c >= n) return false;
        if (grid[r][c] == 1) return true;
        grid[r][c] = 1;
        boolean a = dfs(r+1,c), b = dfs(r-1,c), cc = dfs(r,c+1), d = dfs(r,c-1);
        return a && b && cc && d;
    }
}

File: problems/1254_number_of_closed_islands/solution.py
class Solution:
    def closedIsland(self, grid: list[list[int]]) -> int:
        m, n = len(grid), len(grid[0])
        def dfs(r: int, c: int) -> bool:
            if r < 0 or r >= m or c < 0 or c >= n:
                return False
            if grid[r][c] == 1:
                return True
            grid[r][c] = 1
            a = dfs(r+1, c)
            b = dfs(r-1, c)
            c2 = dfs(r, c+1)
            d = dfs(r, c-1)
            return a and b and c2 and d
        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 0:
                    if dfs(i, j):
                        count += 1
        return count
